File: patterns/behavioral/command.py
"""
Command Pattern - Python

Encapsulates a request as an object, allowing you to parameterize clients
with different requests, queue operations, and support undo operations.

Use cases:
- Undo/Redo functionality
- Queuing requests
- Logging requests
- Remote procedure calls
"""

from abc import ABC, abstractmethod


# Command interface
class Command(ABC):
    @abstractmethod
    def execute(self):
        pass
    
    @abstractmethod
    def undo(self):
        pass


# Receiver
class Light:
    def __init__(self):
        self.is_on = False
    
    def turn_on(self):
        self.is_on = True
        return "Light is ON"
    
    def turn_off(self):
        self.is_on = False
        return "Light is OFF"


# Concrete commands
class TurnOnCommand(Command):
    def __init__(self, light):
        self.light = light
    
    def execute(self):
        return self.light.turn_on()
    
    def undo(self):
        return self.light.turn_off()


class TurnOffCommand(Command):
    def __init__(self, light):
        self.light = light
    
    def execute(self):
        return self.light.turn_off()
    
    def undo(self):
        return self.light.turn_on()


# Invoker
class RemoteControl:
    def __init__(self):
        self.history = []
    
    def execute_command(self, command):
        result = command.execute()
        self.history.append(command)
        return result
    
    def undo(self):
        if self.history:
            command = self.history.pop()
            return command.undo()
        return "Nothing to undo"


# Example usage
if __name__ == "__main__":
    print("=== Command Pattern ===\n")
    
    light = Light()
    turn_on = TurnOnCommand(light)
    turn_off = TurnOffCommand(light)
    
    remote = RemoteControl()
    
    print(remote.execute_command(turn_on))
    print(remote.execute_command(turn_off))
    print("Undo:", remote.undo())
    print("Undo:", remote.undo())

File: patterns/behavioral/command.js
/**
 * Command Pattern - JavaScript
 * 
 * Encapsulates a request as an object, allowing you to parameterize clients
 * with different requests, queue operations, and support undo operations.
 * 
 * Use cases:
 * - Undo/Redo functionality
 * - Queuing requests
 * - Logging requests
 * - Remote procedure calls
 */

// Command interface
class Command {
    execute() {
        throw new Error('Method must be implemented');
    }
    
    undo() {
        throw new Error('Method must be implemented');
    }
}

// Receiver
class Light {
    constructor() {
        this.isOn = false;
    }
    
    turnOn() {
        this.isOn = true;
        return 'Light is ON';
    }
    
    turnOff() {
        this.isOn = false;
        return 'Light is OFF';
    }
}

// Concrete commands
class TurnOnCommand extends Command {
    constructor(light) {
        super();
        this.light = light;
    }
    
    execute() {
        return this.light.turnOn();
    }
    
    undo() {
        return this.light.turnOff();
    }
}

class TurnOffCommand extends Command {
    constructor(light) {
        super();
        this.light = light;
    }
    
    execute() {
        return this.light.turnOff();
    }
    
    undo() {
        return this.light.turnOn();
    }
}

// Invoker
class RemoteControl {
    constructor() {
        this.commands = [];
        this.history = [];
    }
    
    executeCommand(command) {
        const result = command.execute();
        this.history.push(command);
        return result;
    }
    
    undo() {
        if (this.history.length > 0) {
            const command = this.history.pop();
            return command.undo();
        }
        return 'Nothing to undo';
    }
}

// Example usage
console.log('=== Command Pattern ===\n');

const light = new Light();
const turnOn = new TurnOnCommand(light);
const turnOff = new TurnOffCommand(light);

const remote = new RemoteControl();

console.log(remote.executeCommand(turnOn));
console.log(remote.executeCommand(turnOff));
console.log('Undo:', remote.undo());
console.log('Undo:', remote.undo());

module.exports = { Command, Light, TurnOnCommand, TurnOffCommand, RemoteControl };

